File: Task Tracker Java/contest_java/contestRepoDB/src/main/java/com/example/repo/tasks/TaskDBRepository.java
package com.example.repo.tasks;

import com.example.domain.Task;
import com.example.enums.AgeGroup;
import com.example.enums.Type;
import com.example.utils.JDBCutils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

@Component
public class TaskDBRepository implements TaskRepository {
    private JDBCutils jdbCutils;
    private static final Logger logger = LogManager.getLogger();

    public TaskDBRepository(Properties props) {
        logger.info("Initializing Repository with {}", props);
        jdbCutils = new JDBCutils(props);
    }
    @Override
    public List<Task> getAll(){
        logger.traceEntry();
        List<Task> tasks = new ArrayList<>();

        String query = "SELECT * FROM tasks";
        Connection connection = jdbCutils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement(query)){
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                while(resultSet.next()){
                    int taskID = resultSet.getInt("taskid");
                    Type type = Type.valueOf(resultSet.getString("type"));
                    AgeGroup agegroup = AgeGroup.valueOf(resultSet.getString("agegroup"));
                    Task task = new Task(taskID, type, agegroup);
                    tasks.add(task);
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return tasks;
    }

    @Override
    public boolean delete(Task entity) throws IOException {
        logger.traceEntry("deleting tasks {}", entity);

        String query = "DELETE FROM tasks WHERE taskid=?";
        Connection connection = jdbCutils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setInt(1, entity.getId());
            int result = preparedStatement.executeUpdate();
            if(result > 0) {
                return true;
            }
            logger.traceExit("deleted {}", entity);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return false;
    }

    @Override
    public Task update(Task entity) throws IOException {
        logger.traceEntry("updating tasks {}", entity);

        String query = "UPDATE tasks SET type=?, agegroup=? WHERE taskid=?";
        Connection connection = jdbCutils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setString(1, String.valueOf(entity.getType()));
            preparedStatement.setString(2, String.valueOf(entity.getAgeGroup()));
            preparedStatement.setInt(3, entity.getId());
            preparedStatement.executeUpdate();
            logger.traceExit("updated task {}", entity);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return entity;
    }

    private Task findMaxID(){
        String query = "SELECT * FROM tasks where taskid=(SELECT max(taskid) FROM tasks)";
        Connection connection = jdbCutils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    int taskID = resultSet.getInt("taskid");
                    Type type = Type.valueOf(resultSet.getString("type"));
                    AgeGroup agegroup = AgeGroup.valueOf(resultSet.getString("agegroup"));
                    return new Task(taskID, type, agegroup);
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return null;
    }

    @Override
    public Task save(Task entity) throws IOException {
        logger.traceEntry("saving task {}", entity);

        String query = "INSERT INTO tasks(type, agegroup) VALUES (?, ?)";
        Connection connection = jdbCutils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement(query)){
            preparedStatement.setString(1, entity.getType().toString());
            preparedStatement.setString(2, entity.getAgeGroup().toString());
            int result = preparedStatement.executeUpdate();
            if(result > 0){
                return findMaxID();
            }
            logger.trace("Saved {}", result);
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return entity;
    }

    @Override
    public int size() {
        logger.traceEntry("counting tasks");

        String query = "SELECT COUNT(*)  FROM tasks";
        Connection connection = jdbCutils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement(query)){
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if (resultSet.next()){
                    logger.traceExit(resultSet.getInt("count(*)"));
                    return resultSet.getInt("count(*)");
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }

        return 0;
    }

    @Override
    public List<Task> findTaskByAge(AgeGroup age) {
        List<Task> tasks = new ArrayList<>();

        String query = "SELECT * FROM tasks where agegroup=?";
        Connection connection = jdbCutils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setString(1, String.valueOf(age));
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                while(resultSet.next()){
                    int taskID = resultSet.getInt("taskid");
                    Type type = Type.valueOf(resultSet.getString("type"));
                    AgeGroup agegroup = AgeGroup.valueOf(resultSet.getString("agegroup"));
                    Task task = new Task(taskID, type, agegroup);
                    tasks.add(task);
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return tasks;
    }

    @Override
    public List<Task> findTaskByType(Type type) {
        List<Task> tasks = new ArrayList<>();

        String query = "SELECT * FROM tasks where type=?";
        Connection connection = jdbCutils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setString(1, String.valueOf(type));
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                while(resultSet.next()){
                    int taskID = resultSet.getInt("taskid");
                    Type typeD = Type.valueOf(resultSet.getString("type"));
                    AgeGroup agegroup = AgeGroup.valueOf(resultSet.getString("agegroup"));
                    Task task = new Task(taskID, typeD, agegroup);
                    tasks.add(task);
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return tasks;
    }

    @Override
    public Task findById(int id) {
        String query = "SELECT * FROM tasks where taskid=?";
        Connection connection = jdbCutils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setInt(1, id);
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    int taskID = resultSet.getInt("taskid");
                    Type type = Type.valueOf(resultSet.getString("type"));
                    AgeGroup agegroup = AgeGroup.valueOf(resultSet.getString("agegroup"));
                    return new Task(taskID, type, agegroup);
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return null;
    }

    @Override
    public int getTasksByAgeType(AgeGroup age, Type type) {
        String query = "SELECT COUNT(*) FROM tasks t INNER JOIN participantstasks pt ON pt.taskID2 = t.taskid OR pt.taskID1 = t.taskid\n" +
                "                             INNER JOIN participants p on p.participantid = pt.participantID\n" +
                " WHERE type=? AND agegroup=?";
        Connection connection = jdbCutils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setString(1, String.valueOf(type));
            preparedStatement.setString(2, String.valueOf(age));
            ResultSet resultSet = preparedStatement.executeQuery();
            if(resultSet.next()){
                logger.traceExit(resultSet.getInt("count(*)"));
                return resultSet.getInt("count(*)");
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return 0;
    }

    @Override
    public boolean deleteById(int id) {
        logger.traceEntry("deleting task by id {}", id);

        String query = "DELETE FROM tasks WHERE taskid=?";
        Connection connection = jdbCutils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement(query)) {
            preparedStatement.setInt(1, id);
            preparedStatement.executeUpdate();
            logger.traceExit("deleted {}");
        } catch (SQLException ex) {
            logger.error(ex);
            System.err.println("DB Error " + ex);
            throw new RuntimeException(ex);
        }
        return false;
    }
}

File: Task Tracker C#/contest/repo/tasks/TaskDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using contest.domain;
using contest.domain.enums;
using contest.utils;
using log4net;
using Type = contest.domain.enums.Type;

namespace contest.repo.tasks
{
    public class TaskDBRepository : IRepository<int, Task>
    {
        private static readonly ILog logger = LogManager.GetLogger("TaskDBRepository");

        private IDictionary<string, string> Properties;

        public TaskDBRepository(IDictionary<string, string> properties)
        {
            logger.Info("Creating Task DB Repo");
            Properties = properties;
        }

        public IEnumerable<Task> FindAll()
        {
            IDbConnection dbConnection = DButils.GetConnection(Properties);
            IList<Task> tasks = new List<Task>();
            
            using (var comm = dbConnection.CreateCommand())
            {
                logger.InfoFormat("all tasks");
                comm.CommandText = "SELECT id, type, agegroup FROM users";
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int idT = dataR.GetInt32(0);
                        Type type = (Type)Enum.Parse(typeof(Type), dataR.GetString(1));
                        AgeGroup ageGroup = (AgeGroup)Enum.Parse(typeof(AgeGroup), dataR.GetString(2));
                        Task task = new Task(idT, type, ageGroup);
                        tasks.Add(task);
                    }
                }
            }

            return tasks;
        }

        public Task Save(Task e)
        {
            var dbConnection = DButils.GetConnection(Properties);

            using (var comm = dbConnection.CreateCommand())
            {
                logger.InfoFormat("saving task {0}", e);
                comm.CommandText = "INSERT INTO tasks VALUES (@taskid, @type, @agegroup)";
                
                var paramID = comm.CreateParameter();
                paramID.ParameterName = "@taskid";
                paramID.Value = e.ID;
                comm.Parameters.Add(paramID);

                var paramType = comm.CreateParameter();
                paramType.ParameterName = "@type";
                paramType.Value = e.Type.ToString();
                comm.Parameters.Add(paramType);

                var paramAgeGroup = comm.CreateParameter();
                paramAgeGroup.ParameterName = "@agegroup";
                paramAgeGroup.Value = e.AgeGroup.ToString();
                comm.Parameters.Add(paramAgeGroup);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    logger.InfoFormat("Error DB");
                    throw new Exception("No task added!\n");
                }
            }

            return e;
        }

        public bool Delete(Task e)
        {
            throw new NotImplementedException();
        }

        public Task Update(Task e)
        {
            throw new NotImplementedException();
        }
        
        public int Size()
        {
            IDbConnection dbConnection = DButils.GetConnection(Properties);
            
            using (var comm = dbConnection.CreateCommand())
            {
                comm.CommandText = "SELECT count(*) FROM tasks";
                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        logger.InfoFormat("count(*) tasks");
                        return dataR.GetInt32(0);
                    }
                }
            }

            return 0;
        }
    }
}